### PR TITLE
[open-] fix opening a dir with an filetype extension

### PR DESCRIPTION
**Pull request by midichef**
If a directory name looks like a file because of its extension (test.txt, test.tsv), Visidata opens it as if it were a file of that type.

`mkdir test.tsv; vd test.tsv`
```
  File "/home/midichef/.venv/lib/python3.12/site-packages/visidata/loaders/tsv.py", line 89, in iterload
    with self.open_text_source() as fp:
         ^^^^^^^^^^^^^^^^^^^^^^^
[...]
  File "/usr/lib/python3.12/pathlib.py", line 1015, in open
    return io.open(self, mode, buffering, encoding, errors, newline)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
IsADirectoryError: [Errno 21] Is a directory: 'test.tsv'
```

This PR handles that case, including the case where the filename refers to a symlink to a directory.

### CODE REVIEW

**Review by saulpw — APPROVED** [review]
Who is even doing this? :)  Thanks for fixing it!

File: visidata/_open.py [hunk 1]
```diff
@@ -108,6 +108,9 @@ def openPath(vd, p, filetype=None, create=False):
         vd.status('creating blank %s' % (p.given))
         return newfunc(p)
 
+    if p.is_dir():  # also matches symlinks to dirs
+        return vd.open_dir(p)
```
Comment (anjakefala) [inline visidata/_open.py]: This is breaking the vgit tests! They operate on a dir but load them with `--filetype` git. Is there any way to accomodate that usecase?

**Comment by frosencrantz** [comment]
There is also an issue with the filetype guessers and a path that is a directory.  The guessers should not report errors if the path is a directory.

**Comment by midichef** [comment]
@saulpw 
Can you take a careful look at my changes? I don't understand most option handling in visidata, and I don't want to break anything before the 3.1 release.

I've moved two `filetype` assignments into `openPath()` so the logic is all in one place. And I changed the fallback from `''` to `None` to match the default parameter of `openPath()`.

What is `'global'` in `options.getonly`?
https://github.com/saulpw/visidata/blob/d50a1bf9e0f1a65c8cf65165fc2df7dbe7e4c2b3/visidata/_open.py#L151
And how does it differ from vd.options.filetype:
https://github.com/saulpw/visidata/blob/d50a1bf9e0f1a65c8cf65165fc2df7dbe7e4c2b3/visidata/_open.py#L98

Side note, I didn't know we could assign a filetype to a file by its path, like `options.set('filetype', 'csv', '-')`, that's very cool!

**Comment by midichef** [comment]
And what should `guess_*` return to indicate it has no guess for a file: `{}` or `None`? I did `{}` for now.

File: visidata/_open.py (hunk 1 above)
Comment (midichef) [reply visidata/_open.py]: Oops! Okay, I've modified the PR to ignore the extension only for directories, and to accept the filetype specified by the user.

**Comment by midichef** [comment]
Oh, I see the existing behavior of `guess_*` in the case of files that did not match the guess, was to return `None`. I've pushed a change to do that instead of return `{}`.

**Comment by saulpw** [comment]
> What is 'global' in options.getonly?

This is if the option has been set, not as a default or on a specific sheet, but globally by the user (visidatarc or cli or from options sheet).  Whereas options.filetype normally looks at a hierarchy of places for the option to be set, either on the specific sheet, or the sheet class, or by the user, and finally falling back to the default.

File: visidata/loaders/archive.py [hunk 2]
```diff
@@ -10,11 +10,13 @@
 
 @VisiData.api
 def guess_zip(vd, p):
+    if p.is_dir(): return
```
Comment (saulpw) [inline visidata/loaders/archive.py]: This shouldn't be required for guessers; we let it raise the exception and it should be silently ignored.

File: visidata/_open.py [hunk 3]
```diff
@@ -81,6 +81,10 @@ def guess_extension(vd, path):
 def openPath(vd, p, filetype=None, create=False):
     '''Call ``open_<filetype>(p)`` or ``openurl_<p.scheme>(p, filetype)``.  Return constructed but unloaded sheet of appropriate type.
     If True, *create* will return a new, blank **Sheet** if file does not exist.'''
+    # allow user to assign a filetype to a pathname:  options.set('filetype', 'csv', '-')
+    filetype = filetype or vd.options.getonly('filetype', str(p), None)  #1710
+    filetype = filetype or vd.options.getonly('filetype', 'global', None)
```
Comment (saulpw) [inline visidata/_open.py:86]: Why were these lines moved?  It seems like code at line 102/103 should fix the issue by itself.

**Review by saulpw — CHANGES_REQUESTED** [review]
The core fix looks reasonable, can we minimize the diff?

File: visidata/_open.py (hunk 3 above)
Comment (midichef) [reply visidata/_open.py:86]: The code at line 102 would fix it by itself. Lines 85, 86, and 104 are all similar in filling in filetype based on options. The intention was to put them together in one function, to make it easier to trace the logic that sets filetype based on options.

File: visidata/loaders/archive.py (hunk 2 above)
Comment (midichef) [reply visidata/loaders/archive.py]: Okay. My aim was to clean up the `--debug` results. But very few people will ever see those. And I am seeing that visidata style is to take full advantage of ignoring exceptions in such situations, and keep the code lean.

**Review by saulpw — APPROVED** [review]

**Comment by anjakefala** [comment]
@midichef Is your expectation for this PR that now `vd test.tsv -f dir` will work, but `vd test.tsv` might still have a `IsADirectoryError: [Errno 21] Is a directory: 'test.tsv'`?

**Comment by midichef** [comment]
> @midichef Is your expectation for this PR that now `vd test.tsv -f dir` will work, but `vd test.tsv` might still have a `IsADirectoryError: [Errno 21] Is a directory: 'test.tsv'`?

Oops, no. I've force-pushed a fix with this diff as the only change:
```
 def guess_dir(vd, p):
     if p.is_dir():
-        return dict(filetype='dir')
+        return dict(filetype='dir', _likelihood=10)
```
In the case of `vd test.tsv`, there were conflicting guesses for file type. `guess_tsv()` was saying `tsv` with likelihood `3`, and `guess_dir()` was saying `dir` with no likelihood, which is interpreted as `1`. Instead, the new version of `guess_dir()` uses a high priority to override. `10` matches the highest priority currently used in visidata, seen in a few loaders:  `api_airtable.py`, `http.py` and `xlsb.py`.

**Comment by anjakefala** [comment]
Thanks @midichef! =)
